server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -866,6 +887,15 @@ public boolean moveShards() {
                 // can use the cached decision.
                 final var moveDecision = shardMoved ? decideMove(index, shardRouting) : storedShardMovement.moveDecision();
                 if (moveDecision.isDecisionTaken() && moveDecision.cannotRemainAndCanMove()) {
+                    if (notPreferredLogger.isDebugEnabled()) {
+                        logMoveNotPreferred.maybeExecute(
+                            () -> notPreferredLogger.debug(
+                                "Moving shard [{}] from a NOT_PREFERRED allocation, explanation is [{}]",
+                                shardRouting,
+                                moveDecision.getCanRemainDecision().getExplanation()
+                            )
```

Comment:
I think it would be helpful to see where the shard is being moved (moveDecision), not only why it cannot remain. Then we can check whether the shard moved where intended, or got derailed later (either not moved at all, or moved to a different node, perhaps, than the original target).

A canAllocate YES will also give us information about the target node utilization, which might be interesting: `"Shard [%s] in index [%s] can be assigned to node [%s]. The node's utilization would become [%s]"`

Reply:
Oh, maybe this is the problem: https://github.com/elastic/elasticsearch/blob/e0fcab7f2a83173e319f1ca37d016c65404e5fc5/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java#L1019
We lose that information when converting Decision into an AllocationDecision.

Ooph. Okay cool 👍